### PR TITLE
[WIP] Kloud827

### PR DESCRIPTION
**Pull request by mateusz-szczyrzyca**
##### ISSUE TYPE

 - Feature Pull Request
 - New Module Pull Request
 - Bugfix Pull Request
 - Docs Pull Request

##### ANSIBLE VERSION

```

```

##### ADDITIONAL INFORMATION

```

```

### CODE REVIEW

File: contrib/inventory/gce_googleapiclient.py [hunk 1]
```diff
@@ -82,23 +84,28 @@ def get_all_billing_projects(billing_account_name, api_version='v1'):
...
 
 
-def get_all_zones_in_project(project, api_version='v1'):
+def get_all_zones_in_project(project, queue, api_version='v1'):
     zones = []
 
     credentials = GoogleCredentials.get_application_default()
     service = discovery.build('compute', api_version, credentials=credentials)
-
+    print('get_all_zones_in_project, queueid: ', id(queue))
```
Comment (maurofaccenda) [inline contrib/inventory/gce_googleapiclient.py]: remember to remove this when it is done.

File: contrib/inventory/gce_googleapiclient.py [hunk 2]
```diff
@@ -191,7 +199,14 @@ def main(args):
             if zone:
                 zones_list = [zone_name for zone_name in zone]
             else:
-                zones_list = get_all_zones_in_project(project)
+                for i in range(3):
```
Comment (maurofaccenda) [inline contrib/inventory/gce_googleapiclient.py]: create a parameter with a default value for the number of threads. and name `i` as `project_thread` or something better.

File: contrib/inventory/gce_googleapiclient.py [hunk 3]
```diff
@@ -191,7 +199,14 @@ def main(args):
...
+                    thread = Thread(target=get_all_zones_in_project,
+                                    name="Thread1",
```
Comment (maurofaccenda) [inline contrib/inventory/gce_googleapiclient.py]: all threads have the same name?

File: contrib/inventory/gce_googleapiclient.py [hunk 4]
```diff
@@ -191,7 +199,14 @@ def main(args):
...
+                                    args=(project, queue),
+                                    )
```
Comment (maurofaccenda) [inline contrib/inventory/gce_googleapiclient.py]: remove the last comma and the line break `args=(project, queue))`

File: contrib/inventory/gce_googleapiclient.py (hunk 1 above)
Comment (mateusz-szczyrzyca) [reply contrib/inventory/gce_googleapiclient.py]: Yes, print() will be removed in final version

File: contrib/inventory/gce_googleapiclient.py (hunk 2 above)
Comment (mateusz-szczyrzyca) [reply contrib/inventory/gce_googleapiclient.py]: Done

File: contrib/inventory/gce_googleapiclient.py (hunk 3 above)
Comment (mateusz-szczyrzyca) [reply contrib/inventory/gce_googleapiclient.py]: Good point, fixed

File: contrib/inventory/gce_googleapiclient.py (hunk 4 above)
Comment (mateusz-szczyrzyca) [reply contrib/inventory/gce_googleapiclient.py]: Done

File: contrib/inventory/gce_googleapiclient.py [hunk 5]
```diff
@@ -7,8 +7,10 @@
 import json
 import logging as log
 
+from Queue import Queue
```
Comment (bart2) [inline contrib/inventory/gce_googleapiclient.py]: use conditional import to support py2 and py3:
```
if sys.version_info < (3, 0):
    import Queue as queue
else:
    import queue
```

File: contrib/inventory/gce_googleapiclient.py [hunk 6]
```diff
@@ -19,6 +21,9 @@
 
 from oauth2client.client import GoogleCredentials
 
+MAX_ZONES_THREADS = 1
```
Comment (bart2) [inline contrib/inventory/gce_googleapiclient.py:28]: We should have a single thread limit for all work.

File: contrib/inventory/gce_googleapiclient.py [hunk 7]
```diff
@@ -82,30 +87,37 @@ def get_all_billing_projects(billing_account_name, api_version='v1'):
     return project_ids
 
 
-def get_all_zones_in_project(project, api_version='v1'):
+def get_all_zones_in_project(project, queue, api_version='v1'):
```
Comment (bart2) [inline contrib/inventory/gce_googleapiclient.py]: make the "queue" name more descriptive. What is passed through the queue?

File: contrib/inventory/gce_googleapiclient.py [hunk 8]
```diff
@@ -82,30 +87,37 @@ def get_all_billing_projects(billing_account_name, api_version='v1'):
...
     zones = []
 
     credentials = GoogleCredentials.get_application_default()
     service = discovery.build('compute', api_version, credentials=credentials)
-
+    print('get_all_zones_in_project, project: {}, queueid: {} '.format(project, id(queue)))
```
Comment (bart2) [inline contrib/inventory/gce_googleapiclient.py]: When printing string that is formatted, I prefer to use double quotes. Single quotes should be limited to strings that are not interpolated.

File: contrib/inventory/gce_googleapiclient.py [hunk 9]
```diff
@@ -82,30 +87,37 @@ def get_all_billing_projects(billing_account_name, api_version='v1'):
...
     request = service.zones().list(project=project)
     while request is not None:
-        response = request.execute()
+        try:
+            response = request.execute()
+        except HttpError as exc:
+            log.info('Problem with retrieving zones: %s', str(exc))
```
Comment (bart2) [inline contrib/inventory/gce_googleapiclient.py:107]: Add project name to the log message. Also, I think warning level would be more appropriate here.

File: contrib/inventory/gce_googleapiclient.py [hunk 10]
```diff
@@ -82,30 +87,37 @@ def get_all_billing_projects(billing_account_name, api_version='v1'):
...
                                             previous_response=response)
 
-    return zones
+    # return zones
+    queue.put(zones)
+    queue.task_done()
 
 
-def get_instances(project_id, zone, api_version='v1'):
+def get_instances(project_id, zone, queue, api_version='v1'):
```
Comment (bart2) [inline contrib/inventory/gce_googleapiclient.py]: make the "queue" name more descriptive.

File: contrib/inventory/gce_googleapiclient.py [hunk 11]
```diff
@@ -171,6 +185,8 @@ def get_inventory(instances):
 
 
 def main(args):
+    queue_zones = Queue()
```
Comment (bart2) [inline contrib/inventory/gce_googleapiclient.py]: zones_queue is better name IMO

File: contrib/inventory/gce_googleapiclient.py [hunk 12]
```diff
@@ -171,6 +185,8 @@ def get_inventory(instances):
...
+    queue_instances = Queue()
```
Comment (bart2) [inline contrib/inventory/gce_googleapiclient.py]: instances_queue is better name IMO

File: contrib/inventory/gce_googleapiclient.py [hunk 13]
```diff
@@ -171,6 +185,8 @@ def get_inventory(instances):
...
     project = args['--project']
     all_projects = args['--all-projects']
     zone = args['--zone']
```
Comment (bart2) [inline contrib/inventory/gce_googleapiclient.py]: Can't we just assign zone_list here? docopt will return list of zones. Additionally, "zone_names" would be better name I think

File: contrib/inventory/gce_googleapiclient.py [hunk 14]
```diff
@@ -186,18 +202,32 @@ def main(args):
...
-                    instances.append(instance)
+            if not zone:
+
+                for num in range(MAX_ZONES_THREADS):
+                    thread = Thread(target=get_all_zones_in_project,
+                                    name="ThreadZones_" + str(num),
+                                    args=(project, queue_zones),
+                                    )
+                    thread.start()
+                    thread.join()
```
Comment (bart2) [inline contrib/inventory/gce_googleapiclient.py]: Please create empty list and add threads there. Use another for loop to join the threads on that list.

**Review by bart2 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**